Cleanup and addition of os_gethostbyname

k0ekk0ek opened the pull request:
This pull request mainly removes code in the abstraction layer, but also adds a platform independent function to resolve hostnames and throws out the bogus Abstraction and GetTime CMake packages.

* The DDSI module allows the user to specifiy a hostname in the configuration file and it used an textual IP address to binary socket address function to take care of that. I needed to be able to disable resolving hostnames for platforms that do not support domain name resolution (lwIP doesn't by default). Also, by  simply using inet_pton based on a define didn't allow me to properly test, so I opted to move that logic up the stack a little and enable/disable it based on OS_SOCKET_HAS_DNS.
* FindAbstraction.cmake and GetTime.cmake only made things unnecessary complicated in my opinion and I simply moved some of the logic to the CMakeLists.txt for the abstraction layer. Apart from that, on Windows platforms project wide compile definitions were set, I changed that to be private to the abstraction layer. Doing this revealed warnings in other modules. I fixed most of them, but some (those about fopen and sscanf) I opted to suppress.
* Some modules used the FindAbstraction.cmake module directly. This is wrong, they must use the OSAPI target. I've corrected this.
* I removed a lot of definitions that weren't used by anybody anymore to make future ports (I'm working on FreeRTOS) a little easier to make.

I've written tests and ran my builds through scan-build. Please consider pulling my changes.

k0ekk0ek commented:
BTW, it also fixes warnings introduced on Windows introduced with the function renaming pull request this weekend.

eboasson commented on src/os/src/vxworks/os_platform_errno.c:
```diff
@@ -39,44 +39,49 @@ strerrorIf(
+    assert(buf != NULL);
+
+    if ((err = os_errstr(errnum, buf, buflen)) == EINVAL) {
+        /* VxWorks's strerror_r always returns 0 (zero), so the only way to
```
Alternatively, just always fill buf as well as one can, return a pointer to it and don't worry about ``EINVAL`` or ``ERANGE``. That definitely would be my preference.

k0ekk0ek commented:
I agree with most comments.

* Moved the checks for endianness and 32-bit vs 64-bit to CMake. That way it should always be correct for the target a user is compiling for and allowed me to remove some extra cruft from the platform headers.
* Added an os_strlcpy and os_strlcat, which is going a bit overboard perhaps, but I really didn't want to suppress extra warnings on Windows.
* Removed the need for the defines with "aap", "noot", etc by using variables exported by CMake.

Regarding the `os_hostent_t`, I just mimicked the api for gethostbyname because I didn't want to add a whole lot of code just for a simple name lookup. That's also the reason I just removed the SOCK_DGRAM requirement. It's hostname lookup and nothing else. If it's alright with you I'd like to leave it with that for now. I also agree that the os_errstr interface could be simpler. I think that removing os_strerror_r (not os_strerror) eventually (and the need for os_errstr) is best.

macOS is failing, but that's due to a test case expecting a different result. The big-picture should remain more-or-less the same. I'll ping once those problems have been resolved.

k0ekk0ek commented:
Updated pull request to include more changes.
* Replace NN_C99_INLINE and VDDS_INLINE by plain inline.
* Move everything except log_stacktraces from sysdeps.c in DDSI to the abstraction layer.
* Implement os_getrusage for Windows and macOS and use it where getrusage was previously used and where thread cpu times are printed
* Updated gv.default_plist_pp.exec_name to always use the project name and pid so os_progName can be ditched.

k0ekk0ek commented:
The macOS build is failing. Nothing special, the IPv6 test returns fe80::1 which isn't considered to be a loopback address anymore. Working on it.

k0ekk0ek commented:
Updated the IPv6 test for getifaddrs and updated os_sockaddr_is_loopback. Builds and tests should pass on all platforms now.

k0ekk0ek commented on src/core/ddsc/src/dds_init.c:
```diff
@@ -149,17 +149,10 @@ dds_init(dds_domainid_t domain)
-    (void) snprintf(gv.default_plist_pp.exec_name, 32, "%s: %u", DDSC_PROJECT_NAME, gv.default_plist_pp.process_id);
-  }
+  gv.default_plist_pp.exec_name = dds_string_alloc(32);
+  (void) snprintf(gv.default_plist_pp.exec_name, 32, "%s: %u", DDSC_PROJECT_NAME, gv.default_plist_pp.process_id);
```
Agreed. I've added it to my TODO list and will pick it up first thing.

eboasson replied:
Excellent

k0ekk0ek commented on src/core/ddsi/src/q_sockwaitset.c:
```diff
@@ -60,7 +60,7 @@ struct entry {
 {
   int kqueue;
-  os_handle pipe[2];             /* pipe used for triggering */
+  os_socket pipe[2];             /* pipe used for triggering */
```
Agreed. I've chosen os_socket because it was defined as such in the abstraction layer already. Replacing os_socket by os_handle everywhere is a simple search replace and shouldn't be too much work.

k0ekk0ek replied:
The os_socket is the proper abstraction in most places except for the os_sockWaitset. One of the first things I'm going to do after this PR is merged, is push that down to the abstraction layer. Is it ok with you if I take care of it then?

eboasson replied:
Yep, that's fine. Some as with the comment below on another instance of the same phenomenon.

k0ekk0ek commented on src/core/ddsi/src/q_init.c:
```diff
@@ -533,18 +533,6 @@ int rtps_config_prep (struct cfgst *cfgst)
     goto err_config_late_error;
   }
 
-#if ! OS_SOCKET_HAS_IPV6
```
The useIpv6 member was removed when support for raw ethernet was added.

eboasson replied:
Haha, yes, you're right. I renamed it to compat_use_ipv6 and missed this one, but the existence of ``compat_use_ipv6`` has to do with backwards compatibility for a change in the configuration file done prior to the code being opened up as Eclipse Cyclone DDS. So ... what's the point? (Don't bother responding to that: I know the answer 😁)

The kind of misconfiguration is still possible, but the way to detect it certainly should be different. I'll sort that out myself.

k0ekk0ek commented on src/os/src/os_socket.c:
```diff
@@ -215,136 +175,97 @@ os_sockaddrSameSubnet(const os_sockaddr* thisSock,
+            } else {
+                memset(sa, 0, sizeof(os_sockaddr_in));
+                ((os_sockaddr_in *)sa)->sin_family = AF_INET;
+                memcpy(&((os_sockaddr_in *)sa)->sin_addr, &buf, (socklen_t)sizeof(buf));
```
Fixed.
